include/pybind11/eigen.h

```diff
@@ -641,6 +643,343 @@ struct type_caster<Type, enable_if_t<is_eigen_other<Type>::value>> {
+            policy = return_value_policy::copy;
+        }
+        return cast_impl(&src, policy, parent);
+    }
+
+    static handle cast(const Type &src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic
+            || policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::copy;
+        }
+        return cast(&src, policy, parent);
+    }
+
+    static handle cast(Type *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    static handle cast(const Type *src, return_value_policy policy, handle parent) {
+        if (policy == return_value_policy::automatic) {
+            policy = return_value_policy::take_ownership;
+        } else if (policy == return_value_policy::automatic_reference) {
+            policy = return_value_policy::reference;
+        }
+        return cast_impl(src, policy, parent);
+    }
+
+    template <typename C>
+    static handle cast_impl(C *src, return_value_policy policy, handle parent) {
+        bool dec_parent = false;
+        bool writeable = false;
+        switch (policy) {
+            case return_value_policy::move:
+                if (std::is_const<C>::value) {
+                    pybind11_fail("Cannot move from a constant reference");
+                }
+                {
+                    Eigen::aligned_allocator<Type> allocator;
+                    Type *copy = ::new (allocator.allocate(1)) Type(std::move(*src));
+                    src = copy;
+                }
+
+                parent = capsule(src, [](void *ptr) {
+                             Eigen::aligned_allocator<Type> allocator;
+                             Type *copy = (Type *) ptr;
+                             copy->~Type();
+                             allocator.deallocate(copy, 1);
+                         }).release();
+                dec_parent = true;
+                writeable = true;
+                break;
+
+            case return_value_policy::take_ownership:
+                if (std::is_const<C>::value) {
+                    pybind11_fail("Cannot take ownership of a const reference");
+                }
+                parent = capsule(src, [](void *ptr) { delete (Type *) ptr; }).release();
+                dec_parent = true;
+                writeable = true;
+                break;
+
+            case return_value_policy::copy:
+                parent = {};
+                dec_parent = false;
+                writeable = true;
+                break;
+
+            case return_value_policy::reference:
+                parent = none().release();
+                dec_parent = true;
```

Comment:
Instead of this dec_parent logic, just assign the parent to a py::object and release it to the parent handle when you are done. You can also just explicitly check if the handle is a nullptr which seems to be the only case you decref the parent (btw, even decrefing a nullptr is safe in pybind11).

Reply:
Done